refactor: AcceptToMemoryPool

lsilva01 opened the pull request:
This PR refactors AcceptToMemoryPool.

. Remove `AcceptToMemoryPoolWithTime` (after #23173, this function is no longer needed).
. Remove the `CChainParams chainparams` parameter from ATMP as they can be inferred from the current chain state.
. Update the `tx_pool` test with new function signature.

lsilva01 committed:
Remove AcceptToMemoryPoolWithTime

jnewbery commented on src/validation.h:
```diff
@@ -206,19 +206,23 @@ struct PackageMempoolAcceptResult
+ * function and is exposed only for testing. Client code should use ChainstateManager::ProcessTransaction()
  *
- * @param[in]  active_chainstate  Reference to the active chainstate.
+ * @param[in]  chainparams        Tweakable parameters of this instance.
```
I don't think this description "Tweakable parameters of this instance." is very illuminating. Tweakable suggests to me that the user can modify these parameters, but that isn't the case here.

lsilva01 replied:
Changed the description in 034e827 .

jnewbery reviewed (commented):
Concept ACK

I also have a branch that does this (amongst other things) here: https://github.com/jnewbery/bitcoin/tree/2021-09-process-transaction-2. Feel free to take anything you find useful from that branch.

glozow commented on src/validation.h:
```diff
@@ -206,19 +206,23 @@ struct PackageMempoolAcceptResult
-                                       bool bypass_limits, bool test_accept=false) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+MempoolAcceptResult AcceptToMemoryPool(const CChainParams& chainparams, CTxMemPool& pool,
+                                                        CChainState& active_chainstate, const CTransactionRef &tx,
+                                                        int64_t nAcceptTime, bool bypass_limits, bool test_accept)
```
prefer snake case `accept_time`. Doxygen comment should be more descriptice, e.g. that this is the time that the mempool entry will record and will be used to determine when the entry expires.

lsilva01 replied:
Changed `nAcceptTime` to `accept_time` and improved the field description in 034e827

glozow reviewed (commented):
Concept ACK, good cleanup.

lsilva01 commented:
@jnewbery and @glozow  thanks for reviews.
I addressed your suggestions in 034e827.

As these changes are already implemented in another branch, with more other commits, I can close this PR if it's better.

jnewbery commented on src/validation.h:
```diff
@@ -206,12 +206,26 @@ struct PackageMempoolAcceptResult
- * (Try to) add a transaction to the memory pool.
- * @param[in]  bypass_limits   When true, don't enforce mempool fee limits.
- * @param[in]  test_accept     When true, run validation checks but don't submit to mempool.
+ * Try to add a transaction to the mempool with a specified acceptance time. This is an internal
```
"with a specified acceptance time" is redundant with the parameter documentation below.

lsilva01 replied:
Removed in 9a0b0fc

jnewbery commented on src/validation.h:
```diff
@@ -206,12 +206,26 @@ struct PackageMempoolAcceptResult
  */
-MempoolAcceptResult AcceptToMemoryPool(CChainState& active_chainstate, CTxMemPool& pool, const CTransactionRef& tx,
-                                       bool bypass_limits, bool test_accept=false) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+MempoolAcceptResult AcceptToMemoryPool(const CChainParams& chainparams, CTxMemPool& pool,
+                                                        CChainState& active_chainstate, const CTransactionRef &tx,
+                                                        int64_t accept_time, bool bypass_limits, bool test_accept)
```
Please align arguments on new lines after the opening parens:

```suggestion
MempoolAcceptResult AcceptToMemoryPool(const CChainParams& chainparams, CTxMemPool& pool,
                                       CChainState& active_chainstate, const CTransactionRef &tx,
                                       int64_t accept_time, bool bypass_limits, bool test_accept)
```

(We use clang-style `AlignAfterOpenBracket: true`)

lsilva01 replied:
Done in 9a0b0fc

jnewbery commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -230,7 +230,7 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
                it->second.m_result_type == MempoolAcceptResult::ResultType::INVALID);
 
-        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx_pool, tx, bypass_limits));
+        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(Params(), tx_pool, chainstate, tx, GetTime(), bypass_limits, false /* test_accept */));
```
Please use clang-tidy style [bugprone-argument-comment](https://clang.llvm.org/extra/clang-tidy/checks/bugprone-argument-comment.html):

```suggestion
        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(Params(), tx_pool, chainstate, tx, GetTime(), bypass_limits, /*test_accept=*/false));
```

See also https://github.com/bitcoin/bitcoin/pull/22981

(same comment below)

lsilva01 replied:
Done in 9a0b0fc

jnewbery commented on src/validation.cpp:
```diff
@@ -1011,15 +1012,13 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
-                                                      CChainState& active_chainstate,
-                                                      const CTransactionRef &tx, int64_t nAcceptTime,
-                                                      bool bypass_limits, bool test_accept)
+MempoolAcceptResult AcceptToMemoryPool(const CChainParams& chainparams, CTxMemPool& pool,
```
Again, please align new lines with the opening parens.

lsilva01 replied:
Done in 9a0b0fc

jnewbery commented on src/validation.cpp:
```diff
@@ -349,8 +349,9 @@ void CChainState::MaybeUpdateMempoolForReorg(
         if (!fAddToMempool || (*it)->IsCoinBase() ||
-            AcceptToMemoryPool(
-                *this, *m_mempool, *it, true /* bypass_limits */).m_result_type !=
+            AcceptToMemoryPool(Params(),
```
Consider using the global namespace operator `::Params()` to explicitly show that you're calling a global function. I think the long-term goal is to remove these global calls to make it easier to unit test.

Same comment for the ATMP calls below.

lsilva01 replied:
Removed calls to global Params() in ATMP and in the tx_pool test file ( aac6548 )

jnewbery reviewed (commented):
A few more comments inline

> As these changes are already implemented in another branch, with more other commits, I can close this PR if it's better.

No, please continue with this PR! Just feel free to take anything from the other branch that you find useful.

theStack commented:
Concept ACK

theStack commented:
Possible follow-up idea that came to my mind while reviewing this PR: couldn't we simplify the interface of `AcceptToMemoryPool` by getting rid of both the `CChainParams` and `CTxMemPool` parameters, since those can be deduced from the chainstate anyway? The change would look something like
```diff
-MempoolAcceptResult AcceptToMemoryPool(const CChainParams& chainparams, CTxMemPool& pool,
-                                       CChainState& active_chainstate, const CTransactionRef& tx,
+MempoolAcceptResult AcceptToMemoryPool(CChainState& active_chainstate, const CTransactionRef& tx,
                                        int64_t accept_time, bool bypass_limits, bool test_accept)
     EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
+    const CChainParams& chainparams{active_chainstate.m_params};
+    assert(active_chainstate.m_mempool != nullptr);
+    CTxMemPool& pool{*active_chainstate.m_mempool};
     std::vector<COutPoint> coins_to_uncache;
     MemPoolAccept::ATMPArgs args { chainparams, accept_time, bypass_limits, coins_to_uncache,
                                    test_accept, /* m_allow_bip125_replacement */ true };
```
I couldn't think of a case where passing chainparams or a mempool instance that _doesn't_ belong to the active chainstate makes much sense. Drawback though: currently the member `m_mempool` is `protected` and thus not accessible from outside, that one would need to move to the `public` section, like it was done for `m_params` in the last commit of this PR.

lsilva01 commented:
Thanks for the review @theStack . I opened PR #23465 to test your suggestion.
It seems to work fine.

jnewbery commented:
#23173 is merged. Please rebase and I'll review.

jnewbery commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -230,7 +230,8 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
 
-        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx_pool, tx, bypass_limits));
+        CChainState& chainstate{node.chainman->ActiveChainstate()};
+        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate.m_params, tx_pool, chainstate, tx, GetTime(), bypass_limits, /* test_accept= */false));
```
Its seems odd to change this to:

```
        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(::Params(), tx_pool, chainstate, tx, GetTime(), bypass_limits, /* test_accept= */false));
```

in the first commit, and then update it again to use `chainstate.m_params` in the second commit. You can just go straight to using `chainstate.m_params` in the first commit.

lsilva01 replied:
Agreed. Changed it to the first commit.

jnewbery commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -330,7 +331,7 @@ FUZZ_TARGET_INIT(tx_pool, initialize_tx_pool)
         const bool bypass_limits = fuzzed_data_provider.ConsumeBool();
         ::fRequireStandard = fuzzed_data_provider.ConsumeBool();
-        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(node.chainman->ActiveChainstate(), tx_pool, tx, bypass_limits));
+        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate.m_params, tx_pool, node.chainman->ActiveChainstate(), tx, GetTime(), bypass_limits, /* test_accept= */false));
```
```suggestion
        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate.m_params, tx_pool, chainstate, tx, GetTime(), bypass_limits, /* test_accept= */false));
```

lsilva01 replied:
Done.

jnewbery commented on src/validation.h:
```diff
@@ -208,19 +208,26 @@ struct PackageMempoolAcceptResult
+ * Client code should use ChainstateManager::ProcessTransaction()
  *
- * @param[in]  active_chainstate  Reference to the active chainstate.
+ * @param[in]  chainparams        Chain paramenters.
```
This argument could be removed. `AcceptToMemoryPool()` could just take the `m_params` from the `active_chainstate` that's passed in.

lsilva01 replied:
Done.
As this PR is removing the `CChainParams& chainparams` from ATMP (commit 863703c),  I also added a new commit to remove `CTxMemPool& pool` as suggested in https://github.com/bitcoin/bitcoin/pull/23437#issuecomment-962536149.

jnewbery commented on src/validation.h:
```diff
@@ -208,19 +208,26 @@ struct PackageMempoolAcceptResult
+ * @param[in]  active_chainstate  Reference to the active chainstate.
  * @param[in]  tx                 The transaction to submit for mempool acceptance.
+ * @param[in]  accept_time        The timestamp for adding the transaction to the mempool. Usually
+ *                                the current system time, but may be different (eg when loading
+ *                                mempool txs from disk on startup).
```
It's generally not a good idea to document how the function is called in the function comment (since that may change over time, and this comment would be out of date:

```suggestion
 *                                the current system time, but may be different.
```

lsilva01 replied:
Makes sense. Comment removed.

jnewbery commented:
Please update the PR description now that #23173 is merged.

lsilva01 committed:
Remove calls to global Params() in tx_pool test

lsilva01 commented:
Thanks for review @jnewbery . New commits were added to address the suggestions.

jnewbery commented on src/validation.h:
```diff
@@ -208,19 +208,22 @@ struct PackageMempoolAcceptResult
+ * Client code should use ChainstateManager::ProcessTransaction()
  *
  * @param[in]  active_chainstate  Reference to the active chainstate.
- * @param[in]  pool               Reference to the node's mempool.
```
This causes the tx_pool fuzz test to fail since it operates on a mempool that isn't referenced by the `CChainState` object. See https://github.com/bitcoin/bitcoin/pull/23173#discussion_r731895386.

That should probably be fixed in the fuzz test, but it can be left for a future PR. For now, I suggest you just remove the final commit.

MarcoFalke replied:
For reference, this can trivially be fixed by something like:

```cpp
struct MockedChainstate:public CChainstate{
SwapTxPool(CTxMempool*tx_pool){m_mempool=tx_pool;}
}

...

   CTxMempool fuzzed_pool;
   (MockedChainstate*)(&chainstate)->SwapTxPool(&fuzzed_pool);

lsilva01 replied:
Final commit removed.

lsilva01 replied:
I also addressed the `tx_pool.cpp` issue using a different approach in #23465.
The fuzz test performed successfully in local machine and CI.

```diff
- const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx_pool, tx, bypass_limits));
+ CChainState& chainstate{node.chainman->ActiveChainstate()};
+ chainstate.m_mempool = &tx_pool;

+ const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx, GetTime(), bypass_limits, /* test_accept= */false));
```

mjdietzx commented on src/validation.cpp:
```diff
@@ -1079,15 +1079,13 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
 {
     std::vector<COutPoint> coins_to_uncache;
-    auto args = MemPoolAccept::ATMPArgs::SingleAccept(chainparams, nAcceptTime, bypass_limits, coins_to_uncache, test_accept);
+    auto args = MemPoolAccept::ATMPArgs::SingleAccept(chainparams, accept_time, bypass_limits, coins_to_uncache, test_accept);
```
can this be `const`?

lsilva01 replied:
Refers to the `args` variable?

glozow replied:
No - contrary to what the name `args` suggests, it cannot be const because it contains `coins_to_uncache`, which is populated with coins and returned. If `coins_to_uncache` were removed from `ATMPArgs` then it could be const.

As a side thought, I personally think it would be nice/more RAII-like if uncaching would be handled by `MemPoolAccept` (see #21146), which is one approach to removing it from `args`.

lsilva01 replied:
It seems this change also implies changing `MemPoolAccept::AcceptSingleTransaction`. Maybe it can be a follow-up PR.

glozow replied:
Just to clarify, I didn't mean to suggest that you change it - out of scope for this PR. Marking resolved.

mjdietzx reviewed (commented):
Concept and Approach ACK

lsilva01 commented on src/validation.h:
```diff
@@ -208,19 +208,25 @@ struct PackageMempoolAcceptResult
-                                       bool bypass_limits, bool test_accept=false) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+MempoolAcceptResult AcceptToMemoryPool(const CChainParams& chainparams, CTxMemPool& pool,
+                                       CChainState& active_chainstate, const CTransactionRef& tx,
+                                       int64_t accept_time, bool bypass_limits, bool test_accept)
```
Since this PR removes `AcceptToMemoryPoolWithTime` where `bool test_accept` has no default value, I think making it explicit is better.

lsilva01 commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -230,7 +230,7 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
                it->second.m_result_type == MempoolAcceptResult::ResultType::INVALID);
 
-        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx_pool, tx, bypass_limits));
+        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(::Params(), tx_pool, chainstate, tx, GetTime(), bypass_limits, /* test_accept= */false));
```
Thanks. I will fix it.

lsilva01 replied:
Fixed.

stratospher reviewed (commented):
ACK 863703c.

-  [x] f4c45cd
    - `m_params` is made a public member of `CChainParams`.
    - It's also added as an additional parameter to `AcceptToMemoryPool()` along with `accept_time` and call sites updated.
    - `AcceptToMemoryPoolWithTime()` is removed.
-  [x] e1c221b
    - In the tx_pool fuzz test, `::Params()` is replaced with `CChainState`object's `m_params`.
-  [x] 863703c
    - CChainParams's `m_params` is removed as a parameter from `AcceptToMemoryPool()` since it can be derived from the `CChainState` object parameter.

theStack reviewed (approved):
Code-review ACK 863703c63311bf816b2dcdb7edd70be0bf069014 🍵 

Tiny nit: there is an intermediate typo in the doxygen description for the `chainparams` parameter in `AcceptToMemoryPool` (=> s/paramenters/parameters). The first and third commits could also simply be squashed, in order to avoid introducing a temporary parameter (and its doxygen description) and hence also lower the review burden a bit. Not sure if its worth invalidating ACKs now, but I'm happy to re-review in case.

lsilva01 commented:
@theStack  and @stratospher thanks for the review.
The commit were squashed, as suggested.

theStack commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -230,7 +230,8 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
                it->second.m_result_type == MempoolAcceptResult::ResultType::INVALID);
 
-        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx_pool, tx, bypass_limits));
+        CChainState& chainstate{node.chainman->ActiveChainstate()};
```
I think this line could simply be removed, as it is merely shadowing another reference with the same name and content on top of the fuzz test?

lsilva01 replied:
Good catch. Line removed in 4ec8b96.

MarcoFalke replied:
not addressed?

lsilva01 replied:
Yes, it was addressed in 4ec8b96.

MarcoFalke replied:
In the future, it might be good to keep a clean history and not introduce unused stuff in one commit and remove it in the next.

theStack reviewed (approved):
re-ACK 4ec8b96e077a359c7ba877fc4f78d55ac7d5fb22

jnewbery commented:
utACK 4ec8b96e077a359c7ba877fc4f78d55ac7d5fb22

Happy to rereview this once it's rebased.

theStack reviewed (approved):
re-ACK 123f5de8265af07ff1c95ed8078085af6cb589cb

jnewbery commented:
reACK 123f5de8265af07ff1c95ed8078085af6cb589cb

glozow commented:
light code review ACK 123f5de8265af07ff1c95ed8078085af6cb589cb

MarcoFalke commented on src/validation.h:
```diff
@@ -208,19 +208,23 @@ struct PackageMempoolAcceptResult
+ * @param[in]  active_chainstate  Reference to the active chainstate.
  * @param[in]  tx                 The transaction to submit for mempool acceptance.
+ * @param[in]  accept_time        The timestamp for adding the transaction to the mempool. Usually
+ *                                the current system time, but may be different.
```
nit: I don't think we need to explain what the argument means at the call site.
